train.lua

```diff
@@ -32,349 +35,195 @@ cmd:text('Train a character-level language model')
-      net:backward(batchInputs[s], batchGradOutput)
-
---      if debugState.printOutput then
---        print('batchInputs[s]', batchInputs[s])
---        print('batchTarget', batchTarget)
---        print('batchGradOutput', batchGradOutput)
---      end
-
-      seqLoss = seqLoss + batchLoss
---      doBackwardDebug(debugState, batchTarget, batchLoss, batchGradOutput)
-    end
-    net:backwardThroughTime()
-    net:updateParameters(learningRate)
-  elseif opt.backprop == 'noseq' then
-    for s=1,seqLength do
-      net:forget()
-      net:zeroGradParameters()
-      local batchOffset = lua_modulus(epochOffset + (it - 1) * seqLength + (s - 1) + 1, input_len)
-      local targetOffset = lua_modulus(batchOffset + 1, input_len)
-
-      timer_update(timer, 'forward setup')
-      populateBatchInput(batchOffset, debugState, inputStriped, batchInputs[s])
-
-      timer_update(timer, 'backward setup')
-      local batchTarget = makeBatchTarget(targetOffset, debugState, inputStriped)
-
-      timer_update(timer, 'forward run')
-      local batchOutput = net:forward(batchInputs[s])
---      doOutputDebug(debugState, batchOutput)
-
-      timer_update(timer, 'backward run')
-      local batchLoss = crit:forward(batchOutput, batchTarget)
-      local batchGradOutput = crit:backward(batchOutput, batchTarget)
-      net:backward(batchInputs[s], batchGradOutput)
-      if debugState.printOutput then
-        
+    
+    -- exponential learning rate decay
+    if i % loader.ntrain == 0 and opt.lrdecay < 1 then
+      if epoch >= opt.lrdecayafter then
+        local decayfactor = opt.lrdecay
+        optimstate.learningRate = optimstate.learningRate * decayfactor -- decay it
+        print('decayed learning rate by a factor ' .. decayfactor .. ' to ' .. optimstate.learningRate)
       end
-      net:updateParameters(learningRate)
-
-      seqLoss = seqLoss + batchLoss
---      doBackwardDebug(debugState, batchTarget, batchLoss, batchGradOutput)
     end
-  else
-    error('invalid opt.backprop value')
   end
-  timer_update(timer, 'end')
-
-  if debugState.printOutput then
---    timer_dump(timer)
---    print('params:narrow(1,1,seqLength)', params:narrow(1,1,seqLength):reshape(1,seqLength))
---    print('input ', debugState.inputString)
---    print('target', debugState.targetString)
---    print('output', debugState.outputString)
-    print('epoch=' .. epoch, 'it=' .. it .. '/' .. itsPerEpoch, 'seqLoss=' .. seqLoss, 'time=' .. sys.toc())
-  end
-  if globalIt ~= 1 and ((globalIt - 1) % dumpIntervalIts == 0) then
-    local filename = weights_t7:gsub('$DATASET', dataset):gsub('$EPOCH', epoch):gsub('$IT', it)
-    print('filename', filename)
-    local data = {}
-    data.dataset = dataset
-    data.netParams = netParams
-    data.weights = params:float()
-    data.backend = backend
-    data.vocabSize = #ivocab
-    torch.save(outDir .. '/' .. filename, data) 
+  
+  cutorch.synchronize()
```

Comment:
this is CUDA-specific, and therefore wont work correctly for cpu-only or opencl.